distributed/protocol/serialize.py

```diff
@@ -178,6 +220,36 @@ def deserialize(header, frames, deserializers=None):
     --------
     serialize
     """
+    if "is-collection" in header:
+        headers = header["sub-headers"]
+        lengths = header["frame-lengths"]
+        cls = pickle.loads(header["type-serialized"])
```

Comment:
If we wanted to become more paranoid here we  might just include the type name and then map into a dict of known types:

```python
cls = {"tuple": tuple, "list": list, "set": set, "dict": dict}[header["type-serialized"]]
```

This will likely come up the next time someone does a security pass over the code (which happens sometimes) and it would avoid a conversation in the future.

Reply:
(sorry to not say this originally)

Reply:
```suggestion
        cls = {"tuple": tuple, "list": list, "set": set, "dict": dict}[header["type-serialized"]]
```

Reply:
Aren't we already safe? We only enter this block if the type `is` one of those collections, we don't check if they're instances.

Reply:
I mean, we only enter the serialization block for those types.

Reply:
And I also mean, we don't check if they're subclasses (instead of instances, as I said before).

Reply:
Yeah, this is where I'm apologizing for suggesting one thing, having you do that thing, and then suggesting another.

We're safe only if we can assume that the message that we've received was sent by our code as well.  This isn't necessarily true though.  Someone could craft a message on their own and send that.

In general we've made some promises to people that we'll only call `pickle.loads` in very safe situations, and always with an ability to turn it off.  So we try to avoid calling it if we can.

Reply:
I understand now. But in that case, would msgpack be strictly safer? The user could also craft an unsafe msgpack message, no?

Reply:
pickle.loads can call arbitrary Python code. msgpack.loads just produces base Python objects like int/float/str. Certainly someone could craft malformed msgpack messages, but the worst that would happen is that it would cause an exception. People can use pickle.loads to run arbitrary code, suck up data, delete files, and so on.